### PR TITLE
dialects: switch stencil to carry bounds in SSA values types.

### PR DESCRIPTION
**Pull request by PapyChacal**
This enables `stencil.field` and `stencil.temp` to carry bounds information rather than absolute size only.
It switches the operations to not use attributes to encode this information anymore, but their operands/results types.
It potentially decouples analysis and lowering steps, and just feels more natural to me in the MLIR mindset.
As a comparison: layouted memrefs also carry this information in their type, note onloy in the defining operation's attributes, and those bounds are ultimately mapped to memref offsets.

The only exception here is `stencil.store`, as it is the starting point for shape inference. To allow giving the output size, this operation still works with "ub"/"lb" attributes, because typing the operand is the same thing as typing the defining operation's result, so requiring shape inference to be done already..

Starting from here, there's probably a more fitting way to express this on `stencil.store`? But it's clearly not an objective for me ATM, as it wouldn't help working with stencil IR significantly, only readability.

I would recommend going over [tests/filecheck/dialects/stencil/hdiff.mlir](https://github.com/xdslproject/xdsl/compare/main...emilien/stencil-types-bounds#diff-c1ba398d7430fa90ee60e4e973ceb1a3d0af56c20223849f3aa6d8035f5e828d) for a first skim!

### CODE REVIEW

File: xdsl/dialects/experimental/stencil.py [hunk 1]
```diff
@@ -103,9 +102,7 @@ def min(a: IndexAttr, b: IndexAttr | None) -> IndexAttr:
     def max(a: IndexAttr, b: IndexAttr | None) -> IndexAttr:
         if b is None:
             return a
-        return IndexAttr.get(
-            *(max(ae.data, be.data) for ae, be in zip(a.array.data, b.array.data))
-        )
+        return IndexAttr.get(*(max(ae, be) for ae, be in zip(a, b)))
```
Comment (superlopuh) [inline xdsl/dialects/experimental/stencil.py:105]:
```suggestion
        return IndexAttr.get(*map(max, a, b))
```
Supernit, I think this should work

File: xdsl/dialects/experimental/stencil.py [hunk 2]
```diff
@@ -114,42 +111,112 @@ def __iter__(self) -> Iterator[int]:
...
+        else:
+            return tuple(self.bounds.ub - self.bounds.lb)
 
     @staticmethod
     def parse_parameters(parser: Parser) -> list[Attribute]:
+        def parse_interval() -> tuple[int, int] | int:
+            if parser.parse_optional_punctuation("?"):
+                return -1
+            parser.parse_punctuation("[")
+            l = parser.parse_integer(False)
```
Comment (superlopuh) [inline xdsl/dialects/experimental/stencil.py:166]: What does False mean here? Seems like a label would be useful

Comment (PapyChacal) [reply xdsl/dialects/experimental/stencil.py:166]: It means don't allow booleans, I can use a named arg sure !

File: xdsl/dialects/experimental/stencil.py (hunk 1 above)
Comment (PapyChacal) [reply xdsl/dialects/experimental/stencil.py]: Lovely, indeed, thanks!

**Review by AntonLydike — APPROVED** [review]
Really nice work! Awesome!

File: tests/filecheck/dialects/stencil/hdiff_inference.mlir [hunk 3]
```diff
@@ -31,25 +31,25 @@
...
+// CHECK-NEXT:     %3 = "stencil.cast"(%1) : (!stencil.field<?x?x?xf64>) -> !stencil.field<[-4,68]x[-4,68]x[-4,68]xf64>
+// CHECK-NEXT:     %4 = "stencil.load"(%2) : (!stencil.field<[-4,68]x[-4,68]x[-4,68]xf64>) -> !stencil.temp<[-1,65]x[-1,65]x[0,64]xf64>
 // CHECK-NEXT:     %5 = "stencil.apply"(%4) ({
-// CHECK-NEXT:     ^1(%6 : !stencil.temp<?x?x?xf64>):
-// CHECK-NEXT:       %7 = "stencil.access"(%6) {"offset" = #stencil.index<-1, 0, 0>} : (!stencil.temp<?x?x?xf64>) -> f64
-// CHECK-NEXT:       %8 = "stencil.access"(%6) {"offset" = #stencil.index<1, 0, 0>} : (!stencil.temp<?x?x?xf64>) -> f64
-// CHECK-NEXT:       %9 = "stencil.access"(%6) {"offset" = #stencil.index<0, 1, 0>} : (!stencil.temp<?x?x?xf64>) -> f64
-// CHECK-NEXT:       %10 = "stencil.access"(%6) {"offset" = #stencil.index<0, -1, 0>} : (!stencil.temp<?x?x?xf64>) -> f64
-// CHECK-NEXT:       %11 = "stencil.access"(%6) {"offset" = #stencil.index<0, 0, 0>} : (!stencil.temp<?x?x?xf64>) -> f64
+// CHECK-NEXT:     ^1(%6 : !stencil.temp<[-1,65]x[-1,65]x[0,64]xf64>):
```
Comment (georgebisbas) [inline tests/filecheck/dialects/stencil/hdiff_inference.mlir:38]: @PapyChacal

File: tests/filecheck/dialects/stencil/heat_stencil_inference.mlir [hunk 4]
```diff
@@ -185,27 +185,27 @@
...
-// CHECK-NEXT:         %14 = "stencil.access"(%t0_buff) {"offset" = #stencil.index<0, 1, 0>} : (!stencil.temp<?xf32>) -> f32
-// CHECK-NEXT:         %15 = "stencil.access"(%t0_buff) {"offset" = #stencil.index<0, -2, 0>} : (!stencil.temp<?xf32>) -> f32
-// CHECK-NEXT:         %16 = "stencil.access"(%t0_buff) {"offset" = #stencil.index<0, 2, 0>} : (!stencil.temp<?xf32>) -> f32
-// CHECK-NEXT:         %17 = "stencil.access"(%t0_buff) {"offset" = #stencil.index<0, 0, -1>} : (!stencil.temp<?xf32>) -> f32
-// CHECK-NEXT:         %18 = "stencil.access"(%t0_buff) {"offset" = #stencil.index<0, 0, 1>} : (!stencil.temp<?xf32>) -> f32
-// CHECK-NEXT:         %19 = "stencil.access"(%t0_buff) {"offset" = #stencil.index<0, 0, -2>} : (!stencil.temp<?xf32>) -> f32
-// CHECK-NEXT:         %20 = "stencil.access"(%t0_buff) {"offset" = #stencil.index<0, 0, 2>} : (!stencil.temp<?xf32>) -> f32
+// CHECK-NEXT:       ^2(%t0_buff : !stencil.temp<[-2,52]x[-2,82]x[-2,42]xf32>):
+// CHECK-NEXT:         %8 = "stencil.access"(%t0_buff) {"offset" = #stencil.index<0, 0, 0>} : (!stencil.temp<[-2,52]x[-2,82]x[-2,42]xf32>) -> f32
```
Comment (georgebisbas) [inline tests/filecheck/dialects/stencil/heat_stencil_inference.mlir:196]: @PapyChacal

File: xdsl/dialects/experimental/stencil.py [hunk 5]
```diff
@@ -114,42 +113,133 @@ def __iter__(self) -> Iterator[int]:
...
+class StencilBoundsAttr(ParametrizedAttribute):
+    """
+    This attribute represents known bounds over a stencil type.
+    """
+
+    name = "stencil.bounds"
+    lb: ParameterDef[IndexAttr]
+    ub: ParameterDef[IndexAttr]
+
+    def _verify(self):
```
Comment (georgebisbas) [inline xdsl/dialects/experimental/stencil.py:126]: Add failing tests

File: xdsl/transforms/experimental/ConvertStencilToLLMLIR.py [hunk 6]
```diff
@@ -218,14 +219,18 @@ def prepare_apply_body(op: ApplyOp, rewriter: PatternRewriter):
 class ApplyOpToParallel(RewritePattern):
     @op_type_rewrite_pattern
     def match_and_rewrite(self, op: ApplyOp, rewriter: PatternRewriter, /):
-        assert (op.lb is not None) and (op.ub is not None)
+        # if not op.res:
```
Comment (georgebisbas) [inline xdsl/transforms/experimental/ConvertStencilToLLMLIR.py:222]: leftover

File: tests/filecheck/dialects/stencil/heat_stencil_inference.mlir (hunk 4 above)
Comment (PapyChacal) [reply tests/filecheck/dialects/stencil/heat_stencil_inference.mlir:196]: idem

**Pull request merged**